spring-boot-project/spring-boot/src/main/java/org/springframework/boot/webservices/client/WebServiceTemplateBuilder.java

```diff
@@ -0,0 +1,1106 @@
+	 * classpath. Default is {@code true}.
+	 * @param detectWebServiceMessageSender if the {@link WebServiceMessageSender} should
+	 * be detected
+	 * @return a new builder instance
+	 */
+	public WebServiceTemplateBuilder detectWebServiceMessageSender(
+			boolean detectWebServiceMessageSender) {
+		return new WebServiceTemplateBuilder(this.interceptors, this.internalCustomizers,
+				this.customizers, this.webServiceMessageSenderSuppliers,
+				this.webServiceMessageSenderCustomizers, this.marshaller,
+				this.unmarshaller, this.destinationProvider, this.transformerFactoryClass,
+				this.messageFactory, detectWebServiceMessageSender);
+	}
+
+	/**
+	 * Build a new {@link WebServiceTemplate} instance and configure it using this
+	 * builder.
+	 * @return a configured {@link WebServiceTemplate} instance.
+	 * @see #build(Class)
+	 * @see #configure(WebServiceTemplate)
+	 */
+	public WebServiceTemplate build() {
+		return build(WebServiceTemplate.class);
+	}
+
+	/**
+	 * Build a new {@link WebServiceTemplate} instance of the specified type and configure
+	 * it using this builder.
+	 * @param <T> the type of web service template
+	 * @param webServiceTemplateClass the template type to create
+	 * @return a configured {@link WebServiceTemplate} instance.
+	 * @see WebServiceTemplateBuilder#build()
+	 * @see #configure(WebServiceTemplate)
+	 */
+
+	public <T extends WebServiceTemplate> T build(Class<T> webServiceTemplateClass) {
+		Assert.notNull(webServiceTemplateClass,
+				"webServiceTemplateClass must not be null");
+		return configure(BeanUtils.instantiateClass(webServiceTemplateClass));
+	}
+
+	/**
+	 * Configure the provided {@link WebServiceTemplate} instance using this builder.
+	 * @param <T> the type of web service template
+	 * @param webServiceTemplate the {@link WebServiceTemplate} to configure
+	 * @return the web service template instance
+	 * @see #build()
+	 * @see #build(Class)
+	 */
+	public <T extends WebServiceTemplate> T configure(T webServiceTemplate) {
+		Assert.notNull(webServiceTemplate, "webServiceTemplate must not be null");
+
+		configureSenders(webServiceTemplate);
+
+		if (!CollectionUtils.isEmpty(this.internalCustomizers)) {
+			for (WebServiceTemplateCustomizer internalCustomizer : this.internalCustomizers) {
+				internalCustomizer.customize(webServiceTemplate);
+			}
+		}
+
+		if (this.marshaller != null) {
+			webServiceTemplate.setMarshaller(this.marshaller);
+		}
+
+		if (this.unmarshaller != null) {
+			webServiceTemplate.setUnmarshaller(this.unmarshaller);
+		}
+
+		if (this.destinationProvider != null) {
+			webServiceTemplate.setDestinationProvider(this.destinationProvider);
+		}
+
+		if (this.transformerFactoryClass != null) {
+			webServiceTemplate.setTransformerFactoryClass(this.transformerFactoryClass);
+		}
+
+		if (this.messageFactory != null) {
+			webServiceTemplate.setMessageFactory(this.messageFactory);
+		}
+
+		if (!CollectionUtils.isEmpty(this.customizers)) {
```

Comment:
This should be the last step

Reply:
Completely agree, this should be the last step, I did this in favor of `RestTemplateBuilder` algorithm.
```
public <T extends RestTemplate> T configure(T restTemplate) {
		configureRequestFactory(restTemplate);
		if (!CollectionUtils.isEmpty(this.messageConverters)) {
			restTemplate.setMessageConverters(new ArrayList<>(this.messageConverters));
		}
		if (this.uriTemplateHandler != null) {
			restTemplate.setUriTemplateHandler(this.uriTemplateHandler);
		}
		if (this.errorHandler != null) {
			restTemplate.setErrorHandler(this.errorHandler);
		}
		if (this.rootUri != null) {
			RootUriTemplateHandler.addTo(restTemplate, this.rootUri);
		}
		if (this.basicAuthorization != null) {
			restTemplate.getInterceptors().add(this.basicAuthorization);
		}
		if (!CollectionUtils.isEmpty(this.restTemplateCustomizers)) {
			for (RestTemplateCustomizer customizer : this.restTemplateCustomizers) {
				customizer.customize(restTemplate);
			}
		}
		restTemplate.getInterceptors().addAll(this.interceptors);
		return restTemplate;
	}
```

That was weird to me...